### PR TITLE
Add GetLatestErrorMessage feature

### PR DESCRIPTION
**Pull request by wangjuneng**
## What is the purpose of the change

Add GetLatestErrorMessage feature

## Brief changelog
1、Add MQClientErrorContainer to save the latest error message
2、Provide C interface for getting the latest  error message
3、Modify C interface CPullConsumer,CPushConsumer,CPushConsumer  when occur error  use MQClientErrorContainer save error message

## Verifying this change

Follow this checklist to help us incorporate your contribution quickly and easily. Notice, `it would be helpful if you could finish the following 5 checklist(the last one is not necessary)before request the community to review your PR`.

- [x] Make sure there is a [Github issue](https://github.com/apache/rocketmq/issues) filed for the change (usually before you start working on it). Trivial changes like typos do not require a Github issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue. 
- [x] Format the pull request title like `[ISSUE #123] Fix UnknownException when host config not exist`. Each commit in the pull request should have a meaningful subject line and body.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Write necessary unit-test(over 80% coverage) to verify your logic correction, more mock a little better when a cross-module dependency exists.
- [ ] If this contribution is large, please file an [Apache Individual Contributor License Agreement](http://www.apache.org/licenses/#clas).

### CODE REVIEW

**Comment by ShannonDing** [comment]
Maybe a character array defined in commom.c is enough. IMO, we only save the latest error message in C module, the error message can set to the array directly, so the standalone method in CPP does not need. What do you think about it?

**Comment by messense** [comment]
I think this C API should be thread safe?

**Comment by wangjuneng** [comment]
> Maybe a character array defined in commom.c is enough. IMO, we only save the latest error message in C module, the error message can set to the array directly, so the standalone method in CPP does not need. What do you think about it?

I agree with your opinion,Define a  character array in Common.c to save the latest error message.
But I think use global variable will be cause the variable name conflicts.
So Use standalone class to save the latest error message .
there have two benifit.
1.Resolve the variable name conflict.
2.Provide expansion capabilities for the future.

**Comment by ShannonDing** [comment]
> I think this C API should be thread safe?

Maybe we can refer to the processing logic of linux errorno and only keep the last error message. once an error occurs, we need to retrieve the error information immediately. in order to achieve simple logic, the following error information is allowed to cover the old error information.

**Comment by messense** [comment]
@ShannonDing https://linux.die.net/man/3/errno

> errno is defined by the ISO C standard to be a modifiable lvalue of type int, and must not be explicitly declared; errno may be a macro. errno is thread-local; setting it in one thread does not affect its value in any other thread.

I think the error information should be thread-local just like linux errno.

**Comment by ShannonDing** [comment]
@wangjuneng Could you please check the CI result?

File: include/CErrorMessage.h [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+#ifndef __C_CLIENT_ERROR_H__
+#define __C_CLIENT_ERROR_H__
+
+#include "CCommon.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+	 
+	ROCKETMQCLIENT_API const char *GetLatestErrorMessage();  // Return the last error message
```
Comment (jonnxu) [inline include/CErrorMessage.h]: all files in this pr have wrong format, pls format them

**Pull request merged**